lib/Backend/LowerMDShared.cpp

```diff
@@ -5812,6 +5812,165 @@ LowererMD::GenerateCtz(IR::Instr * instr)
+    IR::Opnd* src1 = instr->GetSrc1();
+
+    IR::Opnd * src64 = nullptr;
+    if (src1->IsFloat32())
+    {
+        src64 = IR::RegOpnd::New(TyFloat64, m_func);
+        EmitFloat32ToFloat64(src64, src1, instr);
+    }
+    else
+    {
+        src64 = src1;
+    }
+
+    int min, max;
+    if (instr->GetDst()->IsUInt32())
+    {
+        min = 0;
+        max = INT_MAX; //we will + 2^31 below
+    }
+    else
+    {
+        min = INT_MIN;
+        max = INT_MAX;
+    }
+
+    GenerateCheck(Js::OpCode::UCOMISD, src64, src64, Js::OpCode::JP, throwLabel, instr); //NaN check
+    IR::RegOpnd* limitReg = MaterializeDoubleConstFromInt(min, instr);
+
+    GenerateCheck(Js::OpCode::COMISD, src64, limitReg, Js::OpCode::JB, throwLabel, instr); // < min
+    IR::Opnd * twoTo31Float = MaterializeConstFromBits(_mm_extract_epi32(X86_TWO_31_F4.m128i_value, 0), TyFloat32, instr);
+
+    if (instr->GetDst()->IsUInt32())
+    {
+        limitReg = IR::RegOpnd::New(TyFloat64, m_func);
+        EmitFloat32ToFloat64(limitReg, twoTo31Float, instr);
+        IR::RegOpnd* maxIntDoubleReg = MaterializeDoubleConstFromInt(max, instr);
+        instr->InsertBefore(IR::Instr::New(Js::OpCode::ADDPD, limitReg, limitReg, maxIntDoubleReg, m_func));
+    }
+    else
+    {
+        limitReg = MaterializeDoubleConstFromInt(max, instr);
+    }
+
+    GenerateCheck(Js::OpCode::COMISD, limitReg, src64, Js::OpCode::JAE, conversion, instr); // >= max
+    instr->InsertBefore(throwLabel);
+    IR::Instr *throwInstr = IR::Instr::New(
+        Js::OpCode::RuntimeTypeError,
+        IR::RegOpnd::New(TyMachReg, m_func),
+        IR::IntConstOpnd::New(SCODE_CODE(VBSERR_Overflow), TyInt32, m_func),
+        m_func);
+    instr->InsertBefore(throwInstr);
+    this->m_lowerer->LowerUnaryHelperMem(throwInstr, IR::HelperOp_RuntimeTypeError);
+    //no jump here we aren't coming back
+
+    instr->InsertBefore(conversion);
+}
+
+IR::Opnd*
+LowererMD::TruncLower31BitsFlt(IR::Instr* instr, IR::Opnd* src1, IR::Opnd* dst, IR::Opnd* twoTo31Float)
+{
+    IR::RegOpnd * tmpReg = IR::RegOpnd::New(TyFloat32, m_func);
+    IR::Instr* cmp = IR::Instr::New(Js::OpCode::CMPLTPS, tmpReg, twoTo31Float, src1, m_func);
+
+    instr->InsertBefore(cmp);
+    Legalize(cmp);
+
+    instr->InsertBefore(IR::Instr::New(Js::OpCode::ANDPS, tmpReg, tmpReg, twoTo31Float, m_func));
+    IR::Opnd * result = IR::RegOpnd::New(TyFloat32, m_func);
+    instr->InsertBefore(IR::Instr::New(Js::OpCode::MOVAPS, result, src1, m_func));
+    instr->InsertBefore(IR::Instr::New(Js::OpCode::SUBPS, result, result, tmpReg, m_func));
+    instr->InsertBefore(IR::Instr::New(Js::OpCode::CVTTSS2SI, dst, result, m_func));
+
+    return tmpReg;
+}
+
+IR::Opnd*
+LowererMD::TruncLower31BitsDb(IR::Instr* instr, IR::Opnd* src1, IR::Opnd* dst, IR::Opnd* twoTo31Float)
+{
+    IR::RegOpnd * tmpReg = IR::RegOpnd::New(TyFloat64, m_func);
+    IR::Opnd * result = IR::RegOpnd::New(TyFloat64, m_func);
+    EmitFloat32ToFloat64(result, twoTo31Float, instr);
```

Comment:
we shouldn't add runtime conversion. at jit time just emit load of double value...